### PR TITLE
QOL-7905 Add flask support to CSRF filter

### LINKED ISSUE #28: CSRF not on some POST forms.
@ThrawnCA Continuation of [previous issue](https://github.com/data-govt-nz/ckanext-security/issues/26#issuecomment-562154672).  I originally tried to implement CSRF protection from the ckanext-security extension but with various issues (likely my "user error") ended up jumping to this implementation so I'm moving the conversation here.

**Potential Issue:**
Token is not added to all post forms. Specifically, I've found that the token doesn't get added to the user edit form (user profile).  Can you confirm this on your installation by chance?  If it works there, it's something I've done.

So far I've found [`intercept_csrf()`](https://github.com/qld-gov-au/ckan-ex-qgov/blob/3a84d16e11a1cf7fbe0b6d4a80ecdd19a844cb6d/ckanext/qgov/common/anti_csrf.py#L288) is successfully called when loading the user edit form but it appears that [line 61 (`anti_csrf_render_jinja2`)](https://github.com/qld-gov-au/ckan-ex-qgov/blob/3a84d16e11a1cf7fbe0b6d4a80ecdd19a844cb6d/ckanext/qgov/common/anti_csrf.py#L61) is not called. But when loading another form (e.g. the dataset edit form) it is called as expected.  This happens before `apply_token()` is called, which is where I suspected the issue to be at first.

update:

- organization pages seem fine (e.g. adding, editing, adding users, editing users)
- dataset pages seem fine
- user login, user register, user profile editing, admin config, admin purge is "skipped"

### PR DESCRIPTION
**Pull request by ThrawnCA**
Fixes #28. Thanks @boykoc for the sample code that made this easier.

### CODE REVIEW

File: ckanext/qgov/common/anti_csrf.py [hunk 1]
```diff
@@ -232,11 +236,17 @@ def create_response_token():
     message = "{}/{}/{}".format(timestamp, nonce, username)
     token = "{}!{}".format(get_digest(message), message)
 
-    _set_response_token_cookie(token)
+    # pre-emptively set the token cookie if using Pylons,
+    # otherwise assume the after_app_request hook will do it
+    try:
+        set_response_token_cookie(token, pylons_response)
+    except Exception as e:
+        LOG.warn("Unable to set CSRF token cookie via Pylons: %s", e)
```
Comment (duttonw) [inline ckanext/qgov/common/anti_csrf.py:244]: should this be info not warn. if pylons is not enabled (old ckan) would this not fill the logs?

**Review by duttonw — CHANGES_REQUESTED** [review]
Do we have any tests written that need to be touched up, or new tests added?

**Comment by ThrawnCA** [comment]
Looks like there's a problem with our monkey-patch. The higher-level `render` function that we're now replacing is imported into the plugin toolkit, meaning that other extensions may get the original version. We may need to revert to monkey-patching lower-level functions.

File: ckanext/qgov/common/anti_csrf.py [hunk 2]
```diff
@@ -278,38 +294,37 @@ def _get_post_token():
...
+    request_helpers.scoped_attrs()[TOKEN_FIELD_NAME] = token
+    request_helpers.delete_param(TOKEN_FIELD_NAME)
+    return token
 
 
 def intercept_csrf():
     """ Monkey-patch the core rendering methods to apply our CSRF tokens.
     """
-    base.render_jinja2 = anti_csrf_render_jinja2
+    base.render = anti_csrf_render
```
Comment (duttonw) [inline ckanext/qgov/common/anti_csrf.py:329]: not a fan of this. 
Is there no hooks already in the renderer core that we can't extend off (or see if they can add them {well us to our own version} )

**Comment by duttonw** [comment]
> Looks like there's a problem with our monkey-patch. The higher-level `render` function that we're now replacing is imported into the plugin toolkit, meaning that other extensions may get the original version. We may need to revert to monkey-patching lower-level functions.

should we look at seeing if we can include more 'hooks' into the ckan core renderer if possible to allow csrf to work more neatly on 2.9+ vs 2.8 monkey patches?

**Comment by ThrawnCA** [comment]
> should we look at seeing if we can include more 'hooks' into the ckan core renderer if possible to allow csrf to work more neatly on 2.9+ vs 2.8 monkey patches?

We could bring it up, but they didn't like our CSRF filter's approach when I originally raised it, so they may not be keen on making changes to support it.

File: ckanext/qgov/common/anti_csrf.py (hunk 2 above)
Comment (ThrawnCA) [reply ckanext/qgov/common/anti_csrf.py:329]: Nope, there are none. The entry point (which gets copied to the toolkit and handed out to plugins) is `base.render`, https://github.com/ckan/ckan/blob/master/ckan/lib/base.py#L120
which then calls native Pylons and Flask rendering functions and returns the result.

**Comment by duttonw** [comment]
yep. toolkit has deep link.

https://github.com/qld-gov-au/ckan/blob/qgov-master-2.9.2/ckan/plugins/toolkit.py#L254

still need jinja render hook as its still used (very low usage tho)
https://github.com/qld-gov-au/ckan/blob/qgov-master-2.9.x/ckan/lib/base.py#L117
https://github.com/qld-gov-au/ckan/blob/qgov-master-2.9.x/ckan/lib/base.py#L120

I suggest we add a new interface for renderer something like IApiToken
https://github.com/qld-gov-au/ckan/blob/qgov-master-2.9.x/ckan/plugins/interfaces.py#L2014

Name: IRenderProcessing
Description: 
u'''Return an optional Render Processing object to allow mutation of input variables or output html
        implement the following methods:

pre_render_processing(template_name, extra_vars=None) 
<== allow updates extra_vars only (*pargs, **kwargs is deprecated so i'm not adding them)

post_render_processing(output, template_name, extra_vars=None) 
<== returns output

Update base.py to call get the Interface object, if it is none, we could load in a dummy one or just skip calling the function.

This then allows us to plug in our csr

**Comment by ThrawnCA** [comment]
Raised idea at https://github.com/ckan/ckan/discussions/6036

**Comment by duttonw** [comment]
the way spring does it, is that each form needs to have the csrf item added manually to each form.  This could be a way forward with csrf for ckan, but the owness is on getting the plugins we use to have it plugged in so that they are not failing when csrf is enabled on server response.

**Comment by ThrawnCA** [comment]
> the way spring does it, is that each form needs to have the csrf item added manually to each form. This could be a way forward with csrf for ckan, but the owness is on getting the plugins we use to have it plugged in so that they are not failing when csrf is enabled on server response.

The token is now being applied to Flask responses using the standard `after_app_request` hook.

**Review by duttonw — APPROVED** [review]

**Pull request merged**